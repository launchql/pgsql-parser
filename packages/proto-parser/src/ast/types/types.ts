import { Type, Field } from '@launchql/protobufjs';
import * as t from '@babel/types';
import { createNamedImport, createNamedImportAsSuffix, getFieldName, toSpecialCamelCase } from '../../utils';
import { PgProtoParserOptions } from '../../options';
import { NODE_TYPE } from '../../constants';
import { resolveTypeName } from './utils';

/**
 * Generates import statements for types from a specified source module.
 * Can optionally add a suffix to imported type names (e.g., import { Type as TypeSuffix })
 * Used when types need to be imported into other generated files.
 */
export const generateTypeImports = (types: Type[], source: string, suffix?: string) => {
  return suffix ?
    createNamedImportAsSuffix(types.map(e => e.name), source, suffix) :
    createNamedImport(types.map(e => e.name), source);
};

/**
 * Generates helper factory methods for creating AST node instances.
 * Creates an object with camelCase methods for each type (e.g., selectStmt() for SelectStmt type).
 * Each method accepts optional initial values and returns the type directly.
 * Example output: { selectStmt: (_p?: SelectStmt) => SelectStmt }
 */
export const generateAstHelperMethods = (types: Type[]): t.ExportDefaultDeclaration => {
  const creators = types.map((type: Type) => {
    const typeName = type.name;
    const param = t.identifier('_p');
    param.optional = true;
    param.typeAnnotation = t.tsTypeAnnotation(t.tsTypeReference(t.identifier(typeName)));

    let init: any = [];

    // @ts-ignore
    const fields: Field[] = type.fields;
    const astNodeInit = t.variableDeclaration('const', [
      t.variableDeclarator(
        t.identifier('_j'),
        t.tsAsExpression(
          t.objectExpression(init),
          t.tsTypeReference(t.identifier(typeName))
        )
      )
    ]);

    const setStatements = Object.entries(fields)
      .map(([fName, field]) => {
        const fieldName = getFieldName(field, fName);
        return t.expressionStatement(
          t.callExpression(
            t.memberExpression(t.identifier('_o'), t.identifier('set')),
            [
              t.identifier('_j'),
              t.stringLiteral(fieldName),
              t.optionalMemberExpression(
                t.identifier('_p'),
                t.identifier(fieldName),
                false,
                true
              )
            ]
          )
        );
      });

    // Ensures camel case
    const methodName = toSpecialCamelCase(typeName);

    // Create the method
    const method = t.objectMethod(
      'method',
      t.identifier(methodName),
      [param],
      t.blockStatement([
        astNodeInit,
        ...setStatements,
        t.returnStatement(t.identifier('_j'))
      ])
    );

    // Set return type to just the type
    method.returnType = t.tsTypeAnnotation(t.tsTypeReference(t.identifier(typeName)));
    
    return method;
  });

  return t.exportDefaultDeclaration(t.objectExpression(creators));
}

/**
 * Generates helper factory methods that return AST nodes wrapped in a type-specific object.
 * Creates an object with camelCase methods for each type (e.g., selectStmt() for SelectStmt type).
 * Each method accepts optional initial values and returns { TypeName: TypeInstance }.
 * Example output: { selectStmt: (_p?: SelectStmt) => { SelectStmt: SelectStmt } }
 */
export const generateWrappedAstHelperMethods = (types: Type[]): t.ExportDefaultDeclaration => {
  const creators = types.map((type: Type) => {
    const typeName = type.name;
    const param = t.identifier('_p');
    param.optional = true;
    param.typeAnnotation = t.tsTypeAnnotation(t.tsTypeReference(t.identifier(typeName)));

    let init: any = [];

    // @ts-ignore
    const fields: Field[] = type.fields;
    const astNodeInit = t.variableDeclaration('const', [
      t.variableDeclarator(
        t.identifier('_j'),
        t.tsAsExpression(
          t.objectExpression(init),
          t.tsTypeReference(t.identifier(typeName))
        )
      )
    ]);

    const setStatements = Object.entries(fields)
      .map(([fName, field]) => {
        const fieldName = getFieldName(field, fName);
        return t.expressionStatement(
          t.callExpression(
            t.memberExpression(t.identifier('_o'), t.identifier('set')),
            [
              t.identifier('_j'),
              t.stringLiteral(fieldName),
              t.optionalMemberExpression(
                t.identifier('_p'),
                t.identifier(fieldName),
                false,
                true
              )
            ]
          )
        );
      });

    // Return the wrapped object
    const wrappedReturn = t.returnStatement(
      t.objectExpression([
        t.objectProperty(
          t.identifier(typeName),
          t.identifier('_j')
        )
      ])
    );

    // Ensures camel case
    const methodName = toSpecialCamelCase(typeName);

    // Create the method
    const method = t.objectMethod(
      'method',
      t.identifier(methodName),
      [param],
      t.blockStatement([
        astNodeInit,
        ...setStatements,
        wrappedReturn
      ])
    );

    // Set return type to the wrapped type
    const wrappedType = t.tsTypeLiteral([
      t.tsPropertySignature(
        t.identifier(typeName),
        t.tsTypeAnnotation(t.tsTypeReference(t.identifier(typeName)))
      )
    ]);
    
    method.returnType = t.tsTypeAnnotation(wrappedType);
    
    return method;
  });

  return t.exportDefaultDeclaration(t.objectExpression(creators));
}

/**
 * Generates a TypeScript union type named 'Node' that includes all AST node types.
 * Can generate either a simple union (Type1 | Type2 | ...) or wrapped objects ({ Type1: Type1 } | { Type2: Type2 } | ...)
 * based on the wrappedNodeTypeExport option. This is the main type used to represent any AST node.
 */
export const generateNodeUnionType = (
  options: PgProtoParserOptions,
  types: Type[]
) => {
  if (options.types.wrappedNodeTypeExport) {
    return generateNodeUnionTypeObjectKeys(types);
  }
  const unionTypeNames = types.map(type => t.tsTypeReference(t.identifier(type.name)));

  const unionTypeAlias = t.tsTypeAliasDeclaration(
    t.identifier(NODE_TYPE),
    null,
    t.tsUnionType(unionTypeNames)
  );

  return t.exportNamedDeclaration(unionTypeAlias, []);
};


/**
 * Generates a Node union type where each type is wrapped in an object with the type name as key.
 * Example output: type Node = { SelectStmt: SelectStmt } | { InsertStmt: InsertStmt } | ...
 * This format makes it easier to determine the specific type of a node at runtime.
 */
export const generateNodeUnionTypeObjectKeys = (types: Type[]) => {
  // Mapping types to TypeScript object type annotations
  const unionTypeNames = types.map(type => 
    t.tsTypeLiteral([
      t.tsPropertySignature(
        t.identifier(type.name),
        t.tsTypeAnnotation(t.tsTypeReference(t.identifier(type.name)))
      )
    ])
  );

  // Creating the TypeScript union type
  const unionTypeAlias = t.tsTypeAliasDeclaration(
    t.identifier(NODE_TYPE),
    null,
    t.tsUnionType(unionTypeNames)
  );

  // Exporting the union type declaration
  return t.exportNamedDeclaration(unionTypeAlias, []);
};




// TODO: why is the Field.rule missing in types, but there in the JSON?
interface FieldType extends Field {
  rule: string;
}

const extractTypeFieldsAsTsProperties = (
  type: Type,
  options: PgProtoParserOptions
) => {

  // @ts-ignore
  const fields: FieldType[] = type.fields;
  const properties =
    Object.entries(fields)
      .map(([fieldName, field]) => {
        const type = resolveTypeName(field.type);
        const fieldType = field.rule === 'repeated' ?
          t.tsArrayType(type) :
          type;
        const prop = t.tsPropertySignature(
          t.identifier(getFieldName(field, fieldName)),
          t.tsTypeAnnotation(fieldType)
        );
        prop.optional = options.types.optionalFields;
        return prop;
      });

  return properties;
}

/**
 * Converts a protobuf Type definition into a TypeScript interface declaration.
 * Handles field types, arrays (repeated fields), and optional fields based on options.
 * Example: Type "SelectStmt" with fields becomes: export interface SelectStmt { field1?: Type1; ... }
 */
export const convertTypeToTsInterface = (
  type: Type,
  options: PgProtoParserOptions
) => {
  const typeName = type.name;
  const properties = extractTypeFieldsAsTsProperties(type, options);

  const interfaceDecl = t.tsInterfaceDeclaration(
    t.identifier(typeName),
    null,
    [],
    t.tsInterfaceBody(properties)
  );

  // Wrap the interface declaration in an export statement
  return t.exportNamedDeclaration(interfaceDecl, []);
}


/**
 * Generates import specifiers for types from the configured types source.
 * Used in ast-helpers to import all the type definitions.
 * Example output: import { SelectStmt, InsertStmt, ... } from './types';
 */
export const generateTypeImportSpecifiers = (types: Type[], options: PgProtoParserOptions) => {
  const importSpecifiers = types.map(type =>
    t.importSpecifier(t.identifier(type.name), t.identifier(type.name))
  );

  const importDeclaration = t.importDeclaration(importSpecifiers, t.stringLiteral(options.utils.astHelpers.typesSource));
  return importDeclaration;
}
