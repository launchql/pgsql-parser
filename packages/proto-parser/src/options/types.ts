export type PgProtoParserOptions = PgProtoStoreOptions & {
    parser?: ParseProtoOptions;
};

/**
 * Configuration options for PgProtoStore.
 */
export interface PgProtoStoreOptions {
    // The directory where the generated files will be saved.
    outDir?: string;

    // Types and Enums to skip during processing
    exclude?: string[],

    // Options related to utility functions.
    utils?: {
        enums?: {
            // Whether to include TypeScript utils for enums
            enabled?: boolean;
            // Enums helpers filename
            filename?: string;
            // Whether to generate separate unidirectional functions
            unidirectional?: boolean;
            // Filename for enum to int function
            toIntFilename?: string;
            // Filename for enum to string function
            toStringFilename?: string;
            // Output format: 'switchStatements' or 'nestedObjects'
            outputFormat?: 'switchStatements' | 'nestedObjects';
        }

        // AST helper functions.
        astHelpers?: {
            // Whether to include TypeScript AST builders
            enabled?: boolean;
            // Source file for AST helper types.
            typesSource?: string;
            // Whether to inline nested-obj
            inlineNestedObj?: boolean;
            // if inlined, filename
            nestedObjFile?: string;
            // ASTs helpers filename
            filename?: string;
        }
    };

    // Options related to TypeScript types.
    types?: {
        // Whether to generate TypeScript interfaces for protobuf messages.
        enabled?: boolean;
        // Types filename
        filename?: string;
        // Whether fields in TypeScript interfaces should be optional.
        optionalFields?: boolean;
        // Enums source specifier
        enumsSource?: string;
        // Node Object Keys, e.g. { ParseResult: ParseResult, ... }
        // meant for simpler user cases
        // if a field in an object uses Node, it's a node type
        // if a field in an object references a type, e.g. RangeVar, it's flat
        wrappedNodeTypeExport?: boolean;
    };

    // Options related to runtime schema generation.
    runtimeSchema?: {
        // Whether to generate runtime schema for AST nodes
        enabled?: boolean;
        filename?: string;
        format?: 'json' | 'typescript';
    };

    // Options related to enumeration handling.
    enums?: {
        // Whether to include TypeScript enums.
        enabled?: boolean;
        // Enums filename
        filename?: string;
        // Whether enums in TypeScript are a union type
        enumsAsTypeUnion?: boolean;
        // Options for enum mappings output.
        enumMap?: {
            // Whether to generate enum mapping files.
            enabled?: boolean;
            // Output format: 'json' for plain JSON files, 'ts' for TypeScript exports
            format?: 'json' | 'ts';
            toIntOutFile?: string;
            toStrOutFile?: string;
        };
        // Whether to remove the initial `UNDEFINED` enum entry and adjust subsequent values.
        removeUndefinedAt0?: boolean;
    };

    // Whether to include a header in generated TypeScript files.
    includeHeader?: boolean;
}


export interface ParseProtoOptions {
    keepCase?: boolean;
    alternateCommentMode?: boolean;
    preferTrailingComment?: boolean;
}
