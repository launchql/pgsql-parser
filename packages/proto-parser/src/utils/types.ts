export const isPrimitiveType = (type: string) => {
    switch (type) {
      case 'string':
      case 'double':
      case 'float':
      case 'int32':
      case 'uint32':
      case 'sint32':
      case 'fixed32':
      case 'sfixed32':
      case 'int64':
      case 'uint64':
      case 'sint64':
      case 'fixed64':
      case 'sfixed64':
      case 'bytes':
      case 'bool':
        return true;
      default:
        return false;
    };
  };