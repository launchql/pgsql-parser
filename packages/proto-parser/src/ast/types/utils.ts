import * as t from '@babel/types';
import { isPrimitiveType } from '../../utils';

export const getTSType = (type: string = 'any') => {
  switch (type) {
    case 'string':
      return t.tsStringKeyword();
    case 'double':
    case 'float':
    case 'int32':
    case 'uint32':
    case 'sint32':
    case 'fixed32':
    case 'sfixed32':
      return t.tsNumberKeyword();
    case 'int64':
    case 'uint64':
    case 'sint64':
    case 'fixed64':
    case 'sfixed64':
      return t.tsBigIntKeyword()
    case 'bytes':
      return t.tsTypeReference(t.identifier('Uint8Array'));
    case 'bool':
      return t.tsBooleanKeyword();
    default:
      // For custom types, reference them directly
      return t.tsTypeReference(t.identifier(type));

  };
};

export const resolveTypeName = (type: string) => {
  if (isPrimitiveType(type)) {
    return getTSType(type);
  }
  return t.tsTypeReference(t.identifier(type));
}
