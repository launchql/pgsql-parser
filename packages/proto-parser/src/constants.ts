export const NODE_TYPE = 'Node';
