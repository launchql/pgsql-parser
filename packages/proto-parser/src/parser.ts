import { parse } from '@launchql/protobufjs';
import { readFileSync } from 'fs';
import { ProtoStore } from './store';
import {
  getOptionsWithDefaults,
  PgProtoParserOptions
} from './options';

export class PgProtoParser {
  inFile: string;
  options: PgProtoParserOptions;

  constructor(inFile: string, options?: PgProtoParserOptions) {
    this.inFile = inFile;
    this.options = getOptionsWithDefaults(options);
  }

  private readProtoFile(): string {
    return readFileSync(this.inFile, 'utf-8');
  }

  private parseProto(content: string): any {
    return parse(content, this.options.parser);
  }

  public write(): void {
    const protoContent = this.readProtoFile();
    const ast = this.parseProto(protoContent);
    const store = new ProtoStore(ast.root, this.options);
    store.write();
  }
}
