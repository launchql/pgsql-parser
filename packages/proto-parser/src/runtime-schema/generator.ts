import { Type, Field, Enum, Namespace, ReflectionObject } from '@launchql/protobufjs';
import { NodeSpec, FieldSpec } from './types';

export class RuntimeSchemaGenerator {
  private root: Namespace;
  private nodeTypes: Set<string> = new Set();

  constructor(root: Namespace) {
    this.root = root;
    this.extractNodeTypes();
  }

  private extractNodeTypes(): void {
    const nodeType = this.root.lookupType('Node');
    if (nodeType && nodeType.oneofs && nodeType.oneofs.node) {
      const oneof = nodeType.oneofs.node;
      for (const fieldName of oneof.fieldsArray.map(f => f.name)) {
        const field = nodeType.fields[fieldName];
        if (field && field.type) {
          this.nodeTypes.add(field.type);
        }
      }
    }
  }

  public generateNodeSpecs(): NodeSpec[] {
    const nodeSpecs: NodeSpec[] = [];
    
    const pgQueryNamespace = this.root.nested?.pg_query;
    if (pgQueryNamespace && pgQueryNamespace instanceof Namespace && pgQueryNamespace.nestedArray) {
      pgQueryNamespace.nestedArray.forEach((nested) => {
        if (nested instanceof Type && nested.name !== 'Node') {
          const nodeSpec = this.createNodeSpec(nested);
          if (nodeSpec) {
            nodeSpecs.push(nodeSpec);
          }
        }
      });
    }

    return nodeSpecs.sort((a, b) => a.name.localeCompare(b.name));
  }

  private createNodeSpec(type: Type): NodeSpec | null {
    const fields: FieldSpec[] = [];
    
    Object.values(type.fields).forEach((field) => {
      const fieldSpec = this.createFieldSpec(field);
      if (fieldSpec) {
        fields.push(fieldSpec);
      }
    });

    return {
      name: type.name,
      fields: fields.sort((a, b) => a.name.localeCompare(b.name))
    };
  }

  private createFieldSpec(field: Field): FieldSpec | null {
    const fieldName = field.name;
    const isArray = field.repeated || false;
    const optional = !field.required;
    const fieldType = field.type;

    return {
      name: fieldName,
      type: fieldType,
      isArray,
      optional
    };
  }

  public getNodeTypes(): string[] {
    return Array.from(this.nodeTypes).sort();
  }

  public getNodeTypesCount(): number {
    return this.nodeTypes.size;
  }
}
