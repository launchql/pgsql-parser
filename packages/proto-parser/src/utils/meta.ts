import { toSpecialCamelCase } from './index';
import * as t from '@babel/types';
import { Node } from '@pgsql/types';

export interface FieldSpec {
  name: string;
  type: string;
  isNode: boolean;
  isArray: boolean;
  optional: boolean;
}

export interface NodeSpec {
  name: string;
  isNode: boolean;
  fields: FieldSpec[];
}

/**
 * Converts an AST (Abstract Syntax Tree) representation of a SQL query into
 * a meta-level AST that represents the code to generate the original AST.
 * This meta-level AST can then be used to programmatically construct the AST
 * for similar SQL queries. Essentially, it's an AST that describes how to
 * build an AST, enabling dynamic code generation based on the structure of
 * the original SQL AST. 
 */

// TODO — handle TypeName and SPECIAL_TYPES cases

export function generateTsAstCodeFromPgAst(ast: any): any {
    function createAstNode(functionName: string, properties: any) {
      const args = properties.map(([propKey, propValue]: [string, any]) => {
        return t.objectProperty(t.identifier(propKey), getValueNode(propValue));
      });
      return t.callExpression(
        t.memberExpression(t.identifier('ast'), t.identifier(functionName)),
        [t.objectExpression(args)]
      );
    }
  
    function getValueNode(value: any): t.Expression {
      if (typeof value === 'object') {
        return value === null ? t.nullLiteral() : traverse(value);
      }
      switch (typeof value) {
        case 'boolean':
          return t.booleanLiteral(value);
        case 'number':
          return t.numericLiteral(value);
        case 'string':
          return t.stringLiteral(value);
        default:
          return t.stringLiteral(String(value)); // Fallback for other types
      }
    }
  
    function traverse(node: any): t.Expression {
      if (Array.isArray(node)) {
        return t.arrayExpression(node.map(traverse));
      } else if (node && typeof node === 'object') {
        const entries = Object.entries(node);
        if (entries.length === 0) return t.objectExpression([]);
  
        const [key, value] = entries[0]; // Processing one key-value pair per object
        const functionName = toSpecialCamelCase(key);
        return createAstNode(functionName, Object.entries(value));
      }
  
      return getValueNode(node);
    }
  
    return traverse(ast);
  }

/**
 * Generates TypeScript code from a PostgreSQL AST using runtime schema information.
 * Uses t.ast.* for wrapped nodes and t.nodes.* for structure templates.
 */
export function generateTsAstCodeFromPgAstWithSchema(ast: Node, runtimeSchema: NodeSpec[]): any {
    // Create a map for quick lookup of node specs by name
    const schemaMap = new Map<string, NodeSpec>();
    runtimeSchema.forEach(spec => {
      schemaMap.set(spec.name, spec);
    });

    function createAstNode(functionName: string, properties: any, isWrapped: boolean, nodeSpec?: NodeSpec, forceUnwrapped: boolean = false) {
      const args = properties.map(([propKey, propValue]: [string, any]) => {
        // Check if this field should be treated as a node based on schema
        let processedValue;
        if (nodeSpec) {
          const fieldSpec = nodeSpec.fields.find(f => f.name === propKey);
          if (fieldSpec && fieldSpec.isNode) {
            // This field expects a node
            if (fieldSpec.isArray && Array.isArray(propValue)) {
              // Handle array of nodes - need to check if items should be treated as specific node types
              const itemNodeSpec = schemaMap.get(fieldSpec.type);
              if (itemNodeSpec) {
                processedValue = t.arrayExpression(propValue.map(item => {
                  if (item && typeof item === 'object' && !Array.isArray(item)) {
                    // Check if this is already a wrapped node
                    const entries = Object.entries(item);
                    if (entries.length === 1 && schemaMap.has(entries[0][0])) {
                      // Already wrapped
                      return traverse(item, forceUnwrapped);
                    } else {
                      // Not wrapped - treat as the expected type
                      return createAstNode(
                        toSpecialCamelCase(fieldSpec.type),
                        Object.entries(item).filter(([_, v]) => v !== undefined),
                        itemNodeSpec.isNode && !forceUnwrapped,
                        itemNodeSpec,
                        forceUnwrapped
                      );
                    }
                  }
                  return traverse(item, forceUnwrapped);
                }));
              } else {
                processedValue = t.arrayExpression(propValue.map(item => traverse(item, forceUnwrapped)));
              }
            } else if (!fieldSpec.isArray && propValue && typeof propValue === 'object' && !Array.isArray(propValue)) {
              // Single node - need to detect the actual node type
              // Since the field expects a generic "Node", we need to identify what type of node this is
              processedValue = traverse(propValue, forceUnwrapped);
            } else {
              processedValue = getValueNode(propValue);
            }
          } else {
            processedValue = getValueNode(propValue);
          }
        } else {
          processedValue = getValueNode(propValue);
        }
        
        return t.objectProperty(t.identifier(propKey), processedValue);
      });
      
      // Use t.ast.* for wrapped nodes, t.nodes.* for non-wrapped
      // forceUnwrapped overrides the isWrapped flag
      const namespace = (isWrapped && !forceUnwrapped) ? 'ast' : 'nodes';
      
      return t.callExpression(
        t.memberExpression(
          t.memberExpression(t.identifier('t'), t.identifier(namespace)),
          t.identifier(functionName)
        ),
        [t.objectExpression(args)]
      );
    }
  
    function getValueNode(value: any): t.Expression {
      if (typeof value === 'object') {
        return value === null ? t.nullLiteral() : traverse(value);
      }
      switch (typeof value) {
        case 'boolean':
          return t.booleanLiteral(value);
        case 'number':
          return t.numericLiteral(value);
        case 'string':
          return t.stringLiteral(value);
        default:
          return t.stringLiteral(String(value)); // Fallback for other types
      }
    }
  
    function traverse(node: any, forceUnwrapped: boolean = false): t.Expression {
      if (Array.isArray(node)) {
        return t.arrayExpression(node.map(item => traverse(item, forceUnwrapped)));
      } else if (node && typeof node === 'object') {
        const entries = Object.entries(node);
        if (entries.length === 0) return t.objectExpression([]);
  
        // Check if this is a wrapped node (has single key that matches a node type)
        if (entries.length === 1) {
          const [key, value] = entries[0];
          const nodeSpec = schemaMap.get(key);
          
          if (nodeSpec) {
            // This is a wrapped node
            const functionName = toSpecialCamelCase(key);
            const isWrapped = nodeSpec.isNode && !forceUnwrapped;
            
            // Process the node's fields preserving the original order
            const properties: [string, any][] = [];
            
            if (value && typeof value === 'object') {
              // Process fields in the order they appear in the value object
              Object.entries(value).forEach(([k, v]) => {
                if (v !== undefined) {
                  properties.push([k, v]);
                }
              });
            }
            
            return createAstNode(functionName, properties, isWrapped, nodeSpec, forceUnwrapped);
          }
        }
        
        // Not a wrapped node - but it might be an unwrapped node
        // Try to identify the node type by matching properties
        let identifiedNodeType: string | null = null;
        let identifiedNodeSpec: NodeSpec | null = null;
        let bestMatchCount = 0;
        
        // Check if this object matches any known node type
        const objKeys = Object.keys(node);
        
        for (const [nodeName, nodeSpec] of schemaMap.entries()) {
          // Check if the object has properties that match this node type
          const nodeFields = nodeSpec.fields.map(f => f.name);
          
          // Check if all object keys are valid fields for this node type
          const allPropsValid = objKeys.every(key => nodeFields.includes(key));
          
          // Check if we have at least one field that belongs to this node type
          const hasNodeFields = objKeys.length > 0 && objKeys.some(key => nodeFields.includes(key));
          
          if (allPropsValid && hasNodeFields) {
            // Count matching fields
            const matchingFieldCount = objKeys.filter(key => nodeFields.includes(key)).length;
            
            // Select the node type with the most matching fields
            if (matchingFieldCount > bestMatchCount) {
              bestMatchCount = matchingFieldCount;
              identifiedNodeType = nodeName;
              identifiedNodeSpec = nodeSpec;
            }
          }
        }
        
        if (identifiedNodeType && identifiedNodeSpec) {
          // This is an unwrapped node
          const properties = entries.filter(([_, v]) => v !== undefined);
          // Since this node is not wrapped, we should use t.nodes.* regardless of the schema's isNode value
          return createAstNode(
            toSpecialCamelCase(identifiedNodeType),
            properties,
            false, // isWrapped = false for unwrapped nodes
            identifiedNodeSpec,
            true // forceUnwrapped = true
          );
        }
        
        // Not a node - treat as a plain object
        const properties = entries.filter(([_, v]) => v !== undefined);
        return t.objectExpression(
          properties.map(([k, v]) => t.objectProperty(t.identifier(k), traverse(v, forceUnwrapped)))
        );
      }
  
      return getValueNode(node);
    }
  
    return traverse(ast);
  }
  