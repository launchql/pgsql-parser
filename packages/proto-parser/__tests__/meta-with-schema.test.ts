import * as t from '../test-utils/meta';
import { A_Expr, List, ParseResult, ResTarget, SelectStmt } from '@pgsql/types';
import { generateTsAstCodeFromPgAstWithSchema } from '../src/utils';
import generate from '@babel/generator';
import { runtimeSchema } from '../test-utils/meta/runtime-schema';

describe('AST to AST with runtime schema', () => {
  it('should generate t.ast.* for wrapped nodes and t.nodes.* for non-wrapped nodes', () => {
    // Create a simple AST using the wrapped helpers
    const selectStmt: { SelectStmt: SelectStmt } = t.nodes.selectStmt({
      targetList: [
        t.nodes.resTarget({
          val: t.nodes.columnRef({
            fields: [t.nodes.aStar()]
          })
        })
      ],
      fromClause: [
        t.nodes.rangeVar({
          relname: 'users_table',
          inh: true,
          relpersistence: 'p'
        })
      ],
      limitOption: 'LIMIT_OPTION_DEFAULT',
      op: 'SETOP_NONE'
    });

    // Generate code using the new function with runtime schema
    const astForAst = generateTsAstCodeFromPgAstWithSchema(selectStmt, runtimeSchema);
    const generatedCode = generate(astForAst).code;
    expect(generatedCode).toMatchSnapshot();
  });

  it('should handle mixed wrapped and non-wrapped nodes', () => {
    // Create an AST with both wrapped and non-wrapped nodes
    const testAst: { A_Expr: A_Expr } = t.nodes.aExpr({
      kind: 'AEXPR_OP',
      name: [t.nodes.string({ sval: '=' })],
      lexpr: t.nodes.columnRef({
        fields: [t.nodes.string({ sval: 'id' })]
      }),
      rexpr: t.nodes.aConst({
        ival: t.ast.integer({ ival: 42 })
      })
    });

    const astForAst = generateTsAstCodeFromPgAstWithSchema(testAst, runtimeSchema);
    const generatedCode = generate(astForAst).code;
    expect(generatedCode).toMatchSnapshot();
  });

  it('should handle arrays of nodes', () => {
    const testAst: { List: List } = t.nodes.list({
      items: [
        t.nodes.string({ sval: 'first' }),
        t.nodes.string({ sval: 'second' }),
        t.nodes.string({ sval: 'third' })
      ]
    });

    const astForAst = generateTsAstCodeFromPgAstWithSchema(testAst, runtimeSchema);
    const generatedCode = generate(astForAst).code;
    expect(generatedCode).toMatchSnapshot();
  });

  it('should handle empty objects and null values', () => {
    const testAst: { ResTarget: ResTarget } = t.nodes.resTarget({
      name: null,
      val: t.nodes.columnRef({
        fields: [t.nodes.aStar()]
      })
    });

    const astForAst = generateTsAstCodeFromPgAstWithSchema(testAst, runtimeSchema);
    const generatedCode = generate(astForAst).code;
    expect(generatedCode).toMatchSnapshot();
  });

  it('should use t.nodes.* for non-wrapped nodes', () => {
    // Test that unwrapped nodes (plain objects) are converted to t.nodes.* calls
    const testAst: { ParseResult: ParseResult } = t.nodes.parseResult({
      version: 160001,
      stmts: [
        t.ast.rawStmt({
          stmt: {
            // This is an unwrapped SelectStmt - just a plain object with SelectStmt fields
            targetList: [
              {
                ResTarget: {
                  val: {
                    ColumnRef: {
                      fields: [
                        { A_Star: {} }
                      ]
                    }
                  }
                }
              }
            ],
            limitOption: 'LIMIT_OPTION_DEFAULT',
            op: 'SETOP_NONE'
          }
        })
      ]
    });

    const astForAst = generateTsAstCodeFromPgAstWithSchema(testAst, runtimeSchema);
    const generatedCode = generate(astForAst).code;
    expect(generatedCode).toMatchSnapshot();
  });
});