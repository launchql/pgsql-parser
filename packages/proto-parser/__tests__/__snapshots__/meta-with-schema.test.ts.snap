// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AST to AST with runtime schema should generate t.ast.* for wrapped nodes and t.nodes.* for non-wrapped nodes 1`] = `
"t.ast.selectStmt({
  targetList: [t.ast.resTarget({
    val: t.ast.columnRef({
      fields: [t.ast.aStar({})]
    })
  })],
  fromClause: [t.ast.rangeVar({
    relname: "users_table",
    inh: true,
    relpersistence: "p"
  })],
  limitOption: "LIMIT_OPTION_DEFAULT",
  op: "SETOP_NONE"
})"
`;

exports[`AST to AST with runtime schema should handle arrays of nodes 1`] = `
"t.ast.list({
  items: [t.ast.string({
    sval: "first"
  }), t.ast.string({
    sval: "second"
  }), t.ast.string({
    sval: "third"
  })]
})"
`;

exports[`AST to AST with runtime schema should handle empty objects and null values 1`] = `
"t.ast.resTarget({
  name: null,
  val: t.ast.columnRef({
    fields: [t.ast.aStar({})]
  })
})"
`;

exports[`AST to AST with runtime schema should handle mixed wrapped and non-wrapped nodes 1`] = `
"t.ast.aExpr({
  kind: "AEXPR_OP",
  name: [t.ast.string({
    sval: "="
  })],
  lexpr: t.ast.columnRef({
    fields: [t.ast.string({
      sval: "id"
    })]
  }),
  rexpr: t.ast.aConst({
    ival: t.nodes.aConst({
      ival: 42
    })
  })
})"
`;

exports[`AST to AST with runtime schema should use t.nodes.* for non-wrapped nodes 1`] = `
"t.nodes.parseResult({
  version: 160001,
  stmts: [t.ast.rawStmt({
    stmt: t.nodes.selectStmt({
      targetList: [t.nodes.resTarget({
        val: t.nodes.columnRef({
          fields: [t.nodes.aStar({})]
        })
      })],
      limitOption: "LIMIT_OPTION_DEFAULT",
      op: "SETOP_NONE"
    })
  })]
})"
`;
